### PR TITLE
raft: check pending conf change before campaign

### LINKED ISSUE #12133: Pending conf change is not handled correctly during campaign
After joint consensus is implemented, we add a new conf change type `ConfChangeV2`. But when counting pending conf change, only the old one is counted:

https://github.com/etcd-io/etcd/blob/4c6881ffe4b3bae257c07205c2334f84c0f4577b/raft/raft.go#L1675-L1683

This can lead to two leader at a term problem.

Another problem that has been fixed in raft-rs is that all type of campaign including election timeout and transferring leader should check pending conf change count. But currently etcd/raft only checks pending conf change when campaign is triggered by election timeout. This can also lead to two leader at a term problem. See also tikv/raft-rs#221.

/cc @tbg

### PR DESCRIPTION
**Pull request by BusyJay**
fix #12133.

### CODE REVIEW

File: raft/raft.go [hunk 1]
```diff
@@ -766,6 +766,40 @@ func (r *raft) becomeLeader() {
 	r.logger.Infof("%x became leader at term %d", r.id, r.Term)
 }
 
+func (r *raft) hup(t CampaignType) {
```
Comment (xiang90) [inline raft/raft.go:769]: can you do the cleanup in a separate commit to make it easier for review?

Comment (BusyJay) [reply raft/raft.go:769]: Done, the commit are split into 4, each is summarized as its title.

**Comment by xiang90** [comment]
@BusyJay Create a separate PR for the 3rd commit since it is not relevant to `raft: check pending conf change before campaign` nor `https://github.com/etcd-io/etcd/issues/12133`?  Otherwise, lgtm.

**Comment by BusyJay** [comment]
The commit is removed.

**Comment by xiang90** [comment]
lgtm

File: raft/raft.go [hunk 2]
```diff
@@ -766,6 +766,29 @@ func (r *raft) becomeLeader() {
...
+		return
+	}
+
+	if !r.promotable() {
+		r.logger.Warningf("%x is unpromotable and can not campaign; ignoring MsgHup", r.id)
+		return
+	}
+	ents, err := r.raftLog.slice(r.raftLog.applied+1, r.raftLog.committed+1, noLimit)
+	if err != nil {
+		r.logger.Panicf("unexpected error getting unapplied entries (%v)", err)
```
Comment (cfc4n) [inline raft/raft.go:781]: Is there need `return err` ?

I can't pass `unit tests` with error
```
raft2020/07/23 19:40:17 unexpected error getting unapplied entries (requested index is unavailable due to compaction)
--- FAIL: TestLeaderTransferAfterSnapshot (0.00s)
panic: unexpected error getting unapplied entries (requested index is unavailable due to compaction) [recovered]
	panic: unexpected error getting unapplied entries (requested index is unavailable due to compaction)

goroutine 213 [running]:
testing.tRunner.func1.1(0xba71e0, 0xc000055bc0)
	/usr/local/go1.14.4/go/src/testing/testing.go:940 +0x421
testing.tRunner.func1(0xc000477680)
	/usr/local/go1.14.4/go/src/testing/testing.go:943 +0x600
panic(0xba71e0, 0xc000055bc0)
	/usr/local/go1.14.4/go/src/runtime/panic.go:975 +0x3e3
```

Comment (cfc4n) [reply raft/raft.go:781]: @BusyJay Can you review this comment , It block my PR test ....

Comment (BusyJay) [reply raft/raft.go:781]: This is an error that needs to be fixed by the removed commit. I will send it as a separate PR.

Comment (BusyJay) [reply raft/raft.go:781]: PR is sent as #12163.
